docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,604 @@
+
+Fleet is an automation layer on top of Device API, the DeviceTemplate is a mechanism to control drift
+and enforce policy across devices in a fleet.
+
+A DeviceTemplate is a template for a device spec, it's a way to define group specific configuration.
+
+[We are proposing](https://github.com/flightctl/flightctl/pull/248) extension mechanisms to this API
+to allow for more complex configurations where device labels or name would influence the
+configuration generated to the device.
+
+#### `/fleet/{name}` (user facing)
+This endpoint provides read/write access to fleets.
+
+Full example of fleet:
+
+```yaml
+apiVersion: v1alpha1
+kind: Fleet
+metadata:
+  creationTimestamp: "2024-04-30T14:06:17Z"
+  generation: 1
+  labels: {}
+  name: optical-inspector-production-fleet
+  owner: ResourceSync/basic-nginx-r3sourcesync
+spec:
+  selector:
+    matchLabels:
+      device: optical-inspector
+  template:
+    metadata:
+      generation: 1
+    spec:
+      config:
+      - configType: InlineConfigProviderSpec
+        inline:
+          ignition:
+            version: 3.4.0
+          storage:
+            files:
+            - contents:
+                source: data:,This%20system%20is%20managed%20by%20flightctl.%0A
+              mode: 422
+              overwrite: true
+              path: /etc/motd
+        name: motd-update
+      - configType: GitConfigProviderSpec
+        gitRef:
+          path: /basic-nginx-demo/configuration
+          repository: basic-nginx-d3mo
+          targetRevision: demo
+        name: microshift-manifests
+      - configType: GitConfigProviderSpec
+        gitRef:
+          path: /configuration
+          repository: ricky-super-secrets
+          targetRevision: main
+        name: ricky-super-secrets
+      os:
+        image: quay.io/flightctl/flightctl-agent-basic-nginx:latest
+      systemd:
+        matchPatterns:
+        - microshift.service
+        - crio.service
+        - flightctl-agent.service
+status:
+  conditions:
+  - lastTransitionTime: "2024-05-09T14:20:21Z"
+    status: "False"
+    type: OverlappingSelectors
+```
+
+
+#### Expected behavior
+
+> [!NOTE]
+> This section dives into the implementation details of the fleet controller, the `TemplateVersion`
+> is an internal object that is used to track changes to fleet spec device templates and source references.
+> It is useful to provide an audit trail of the changes to the fleet templates and sources.
+> It's also useful to enable implementations where offline devices can traverse a
+> history of configurations to reach the current state.
+
+When `fleet.spec.template` is updated the Fleet deployment changes, a copy of the device template
+is stored as a `TemplateVersion` object. This object freezes a copy of the template and also creates
+a list of external references (like git refs, image tags, etc..) that are floating tags by scanning
+all fleet devices, these floating tags are resolved to stable hashes and stored in (or along)
```

Comment:
Be more explicit about secret handling since we can't get older versions from k8s

Reply:
Ok we agreed not to be more explicit here, but will be an implementation detail.